### PR TITLE
C++: Do not forward-declare dependencies in generated .h files

### PR DESCRIPTION
**Pull request by ngg**
Those classes and default instances are already forward-declared in included header files.
It solves #2965 where multiple forward-declarations had different dll linkage inside the protobuf compiler.

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

File: src/google/protobuf/compiler/cpp/cpp_file.cc [hunk 1]
```diff
@@ -824,19 +824,11 @@ void FileGenerator::GenerateNamespaceClosers(io::Printer* printer) {
 
 void FileGenerator::GenerateForwardDeclarations(io::Printer* printer) {
```
Comment (xfxyjwf) [inline src/google/protobuf/compiler/cpp/cpp_file.cc]: Where is this GenerateForwardDeclarations() method called? Can you remove that callsite instead?

File: src/google/protobuf/any.pb.h [hunk 2]
```diff
@@ -167,7 +167,7 @@ class LIBPROTOBUF_EXPORT Any : public ::google::protobuf::Message /* @@protoc_in
   ::google::protobuf::internal::ArenaStringPtr value_;
   mutable int _cached_size_;
   ::google::protobuf::internal::AnyMetadata _any_metadata_;
-  friend struct LIBPROTOBUF_EXPORT protobuf_google_2fprotobuf_2fany_2eproto::TableStruct;
+  friend struct protobuf_google_2fprotobuf_2fany_2eproto::TableStruct;
```
Comment (xfxyjwf) [inline src/google/protobuf/any.pb.h]: This change seems suspiciously. I don't see any related protoc change.

**Comment by xfxyjwf** [comment]
ok to test

File: src/google/protobuf/any.pb.h (hunk 2 above)
Comment (ngg) [inline src/google/protobuf/any.pb.h]: the suspicuous change is due to https://github.com/google/protobuf/commit/b4b855cde4197de605d463630af9f0ea89583bca
generate_descriptor_proto.sh was not run after this change

File: src/google/protobuf/compiler/cpp/cpp_file.cc (hunk 1 above)
Comment (ngg) [inline src/google/protobuf/compiler/cpp/cpp_file.cc]: It is called from FileGenerator::GenerateHeader().
That call is still neccesary for the types that are declared in the main protobuf file.
Only the dependencies are unneccesary. This is consistent with the generated .pb.cc file where the *\_default\_instance\_ variables are not written for dependencies either.

**Comment by amerry** [comment]
Is this something you would consider merging into the 3.3 branch? This is the sort of warning that it seems dangerous to just ignore.

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by ngg** [comment]
rebased to current master

**Pull request merged**